src/codegen/compile.ml

```diff
@@ -6154,18 +6171,84 @@ module RTS_Exports = struct
    *)
   let system_exports env =
 
-    (* Value constructors *)
+    (* Helper functions *)
+
+    let export_fun env name f =
+      let fi = E.add_fun env name f in
+      E.add_export env (nr {
+        name = Lib.Utf8.decode name;
+        edesc = nr (FuncExport (nr fi))
+      });
+    in
+
+    (* Value conversions *)
 
-    let int_from_i32_fi = E.add_fun env "int_from_i32" (
+    export_fun env "int_from_i32" (
       Func.of_body env ["v", I32Type] [I32Type] (fun env ->
-        let get_v = G.i (LocalGet (nr 0l)) in
-        get_v ^^ BigNum.from_signed_word32 env
-      )
-    ) in
-    E.add_export env (nr {
-      name = Lib.Utf8.decode "int_from_i32";
-      edesc = nr (FuncExport (nr int_from_i32_fi))
-    });
+        G.i (LocalGet (nr 0l)) ^^
+        BigNum.from_signed_word32 env));
+
+    export_fun env "int8_from_i8" (
+      Func.of_body env ["v", I32Type] [I32Type] (fun env ->
+        G.i (LocalGet (nr 0l)) ^^
+        TaggedSmallWord.msb_adjust Type.Int8 ^^
+        TaggedSmallWord.tag env Type.Int8));
+    export_fun env "nat8_from_u8" (
+      Func.of_body env ["v", I32Type] [I32Type] (fun env ->
+        G.i (LocalGet (nr 0l)) ^^
+        TaggedSmallWord.msb_adjust Type.Nat8 ^^
+        TaggedSmallWord.tag env Type.Nat8));
+    export_fun env "int16_from_i16" (
+      Func.of_body env ["v", I32Type] [I32Type] (fun env ->
+        G.i (LocalGet (nr 0l)) ^^
+        TaggedSmallWord.msb_adjust Type.Int16 ^^
+        TaggedSmallWord.tag env Type.Int16));
+    export_fun env "nat16_from_u16" (
+      Func.of_body env ["v", I32Type] [I32Type] (fun env ->
+        G.i (LocalGet (nr 0l)) ^^
+        TaggedSmallWord.msb_adjust Type.Nat16 ^^
+        TaggedSmallWord.tag env Type.Nat16));
+    export_fun env "int32_from_i32" (
+      Func.of_body env ["v", I32Type] [I32Type] (fun env ->
+        G.i (LocalGet (nr 0l)) ^^ BoxedSmallWord.box env Type.Int32));
+    export_fun env "nat32_from_u32" (
+      Func.of_body env ["v", I32Type] [I32Type] (fun env ->
+        G.i (LocalGet (nr 0l)) ^^ BoxedSmallWord.box env Type.Nat32));
+    export_fun env "int64_from_i64" (
+      Func.of_body env ["v", I64Type] [I32Type] (fun env ->
+        G.i (LocalGet (nr 0l)) ^^ BoxedWord64.box env Type.Int64));
+    export_fun env "nat64_from_u64" (
+      Func.of_body env ["v", I64Type] [I32Type] (fun env ->
+        G.i (LocalGet (nr 0l)) ^^ BoxedWord64.box env Type.Nat64));
+
+    export_fun env "i8_from_int8" (
+      Func.of_body env ["v", I32Type] [I32Type] (fun env ->
+        G.i (LocalGet (nr 0l)) ^^ TaggedSmallWord.lsb_adjust Type.Int8));
```

Comment:
Probably, `TaggedSmallWord.untag` should be applied too before `lsb_adjust`, e.g. to sanity check that the tags are correct.